cvmfs/swissknife_list_reflog.cc

```diff
@@ -0,0 +1,199 @@
+          break;
+        default:
+          LogCvmfs(kLogCvmfs, kLogStderr, "Failed to fetch manifest: %s",
+                    Code2Ascii(failure));
+          return 1;
+      }
+    }
+    success = Run(&object_fetcher, repo_name, output_path, reflog_hash);
+  } else {
+    LocalObjectFetcher<> object_fetcher(repo_url, temp_directory);
+    success = Run(&object_fetcher, repo_name, output_path, reflog_hash);
+  }
+
+  return (success) ? 0 : 1;
+}
+
+template <class ObjectFetcherT>
+bool CommandListReflog::Run(ObjectFetcherT *object_fetcher, string repo_name,
+                            string output_path, shash::Any reflog_hash)
+{
+  typename ObjectFetcherT::ReflogTN *reflog;
+  reflog = FetchReflog(object_fetcher, repo_name, reflog_hash);
+
+  shash::Any null_hash = shash::Any(reflog_hash.algorithm);
+  objects_ = new SmallHashDynamic<shash::Any, bool>;
+  objects_->Init(1024, null_hash, hasher);
+
+  // Traverse through catalogs and regular objects
+  vector<shash::Any> catalogs;
+  if (NULL == reflog || !reflog->List(SqlReflog::kRefCatalog, &catalogs)) {
+    LogCvmfs(kLogCvmfs, kLogStderr, "Failed to list catalog reference log");
+    return false;
+  }
+  typename CatalogTraversal<ObjectFetcherT>::Parameters traversal_params;
+  traversal_params.object_fetcher = object_fetcher;
+  CatalogTraversal<ObjectFetcherT> traversal(traversal_params);
+  traversal.RegisterListener(
+    &swissknife::CommandListReflog::CatalogCallback,
+    this);
+  bool success = true;
+  vector<shash::Any>::iterator i = catalogs.begin();
+  const vector<shash::Any>::const_iterator iend = catalogs.end();
+  for (; i != iend && success; i++) {
+    success &= traversal.TraverseRevision(*i,
+                CatalogTraversal<ObjectFetcherT>::kBreadthFirstTraversal);
+  }
+
+  if (!success) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Catalog traversal aborted due to an error");
+    return false;
+  }
+
+  // Add history, certificate, metainfo objects from reflog
+  vector<shash::Any> histories, certificates, metainfos;
+  if (!reflog->List(SqlReflog::kRefHistory, &histories)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch history objects from reflog");
+    return false;
+  }
+  if (!reflog->List(SqlReflog::kRefCertificate, &certificates)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch certificate objects from reflog");
+    return false;
+  }
+  if (!reflog->List(SqlReflog::kRefMetainfo, &metainfos)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+             "Failed to fetch metainfo objects from reflog");
+    return false;
+  }
+  InsertObjects(histories);
+  InsertObjects(certificates);
+  InsertObjects(metainfos);
+
+  // Clean up reflog file
+  delete reflog;
+
+  LogCvmfs(kLogCvmfs, kLogStderr, "Number of objects: %lu", objects_->size());
+
+  if (output_path == "") {
+    DumpObjects(stdout);
+  } else {
+    int fd = open(output_path.c_str(), O_WRONLY | O_CREAT, 0644);
```

Comment:
we could directly use `fopen`

Reply:
I agree. Although I did it this way because the file can be created with specific permissions. `fopen` on my system creates file with 0664 permissions and I'm not sure how does it behave on other platforms. Do you still want to use `fopen` instead?